### PR TITLE
FPGA: Add code sample `annotated_class_clean_coding`

### PR DESCRIPTION
**Pull request by wangdi4**
# Add code sample `annotated_class_clean_coding`

## Test run:
Has verified the code sample passed by running it locally. 

**Note**: currently we are not able to verify it by running regtest because the current sycl resource for the 2024.1 doesn't contain the annotated USM allocation feature that this tutorial depends on.  **If we can't get the regtest to pass before release then we pull the sample.**

## Description

Add a new FPGA code sample `annotated_class_clean_coding` (original PR: https://github.com/wangdi4/oneAPI-samples/pull/13)

## Checklist
Administrative
- [x] Review sample design with the appropriate [Domain Expert](https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts): <insert Name Here>
- [ ] If you have any new dependencies/binaries, inform the oneAPI Code Samples Project Manager

Code Development
- [ ] Implement coding guidelines and ensure code quality. [see wiki for details](https://github.com/oneapi-src/oneAPI-samples/wiki/General-Code-Guidelines)
- [x] Adhere to readme template 
- [x] Enforce format via clang-format config file
- [x] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [ ] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [x] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com

Security and Legal
- [ ] OSPDT Approval (see Project Manager for assistance)
- [x] Compile using the following compiler flags and fix any warnings, the falgs are: "/Wall -Wformat-security -Werror=format-security"
- [ ] Bandit Scans (Python only)
- [ ] Virus scan

Review
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth(@tomlenth) and/or Project Manager
- [ ] Tested using Dev Cloud when applicable

### CODE REVIEW

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/annotated_class_clean_coding/src/annotated_coding.cpp [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+  void operator()() const {
+#pragma unroll 8
+    for (int i = 0; i < size; i++) {
+      a[i] *= 2;
+    }
+  }
+};
+
+bool checkResult(int *arr, int size) {
```
Comment (whitepau) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/annotated_class_clean_coding/src/annotated_coding.cpp:33]:
```suggestion
bool CheckResult(int *arr, int size) {
```
Code sample code standards request that function names be in `PascalCase()`

(I forgot about this until today, I should have mentioned this sooner)

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/annotated_class_clean_coding/src/annotated_coding.cpp [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    if (arr[i] != golden) {
+      std::cout << "ERROR! At index: " << i << " , expected: " << golden
+                << " , found: " << arr[i] << "\n";
+      passed = false;
+    }
+  }
+  return passed;
+}
+
+bool runWithUsmMalloc(sycl::queue &q) {
```
Comment (whitepau) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/annotated_class_clean_coding/src/annotated_coding.cpp:46]:
```suggestion
bool RunWithUsmMalloc(sycl::queue &q) {
```

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/annotated_class_clean_coding/src/annotated_coding.cpp [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  // Create and initialize the host arrays
+  constexpr int kN = 8;
+  std::cout << "using fpga_tools::alloc_annotated to allocate a block of "
+               "shared memory\n";
+
+  // The "alloc_annotated" function extracts the buffer location and alignment
+  // properties from the type alias annotated_arg_t, rather than forcing you to
+  // explicitly define them in your code. This ensures the properties of the
+  // returned pointer match with the annotations on the kernel arguments.
+  annotated_arg_t array_a = fpga_tools::alloc_annotated<annotated_arg_t>(kN, q);
```
Comment (whitepau) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/annotated_class_clean_coding/src/annotated_coding.cpp:83]:
```suggestion
  annotated_arg_t array_a = fpga_tools::alloc_annotated<annotated_arg_t>(kN, q);
```
variable names need to be `snake_case` (looks good to me)

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/annotated_class_clean_coding/src/annotated_coding.cpp [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+  q.single_task(MyIP{array_a, kN}).wait();
+  bool passed = checkResult(array_a, kN);
+  if (passed) {
+  }
+  sycl::free(array_a, q);
+  return passed;
+}
+
+bool runWithAnnotatedAlloc(sycl::queue &q) {
```
Comment (whitepau) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/annotated_class_clean_coding/src/annotated_coding.cpp:73]:
```suggestion
bool RunWithAnnotatedAlloc(sycl::queue &q) {
```

**Comment by wangdi4** [comment]
@pmpeter1 @tomlenth could you help review this PR?

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/annotated_class_clean_coding/src/annotated_class_util.hpp [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+  static constexpr size_t alignment =
+      sycl::ext::oneapi::experimental::detail::GetAlignFromPropList<
+          all_properties>::value;
+  using properties = typename remove_align_from<all_properties>::type;
+};
+
+// Wrapper function that allocates USM host memory with compile-time properties
+// and returns annotated_ptr
+template <typename T>
+T alloc_annotated(size_t count, const sycl::queue &syclQueue,
```
Comment (pmpeter1) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/annotated_class_clean_coding/src/annotated_class_util.hpp:172]: The use of camelCase for syclQueue and usmKind seems to violate a consistent naming convention for this file.

**Review by pmpeter1 — CHANGES_REQUESTED** [review]

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/annotated_class_clean_coding/src/annotated_class_util.hpp (hunk 5 above)
Comment (wangdi4) [reply DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/annotated_class_clean_coding/src/annotated_class_util.hpp]: The variable names have been fixed.

**Review by pmpeter1 — APPROVED** [review]

**Comment by tomlenth** [comment]
looks fine. Thank you.

**Review by whitepau — APPROVED** [review]

**Pull request merged**